mcswupd/main.go

```diff
@@ -0,0 +1,298 @@
+
+	return nil
+}
+
+func failIfErr(err error) {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "ERROR: %s\n", err)
+		os.Exit(1)
+	}
+}
+
+func buildBundles(b *builder.Builder) error {
+	if _, err := os.Stat(b.Config.Builder.Cert); os.IsNotExist(err) {
+		fmt.Println("Generating certificate for signature validation...")
+		privkey, err := helpers.CreateKeyPair()
+		if err != nil {
+			return errors.Wrap(err, "Error generating OpenSSL keypair")
+		}
+		template := helpers.CreateCertTemplate()
+		err = b.BuildBundles(template, privkey, false)
+		if err != nil {
+			return errors.Wrap(err, "Error building bundles")
+		}
+	} else {
+		err := b.BuildBundles(nil, nil, false)
+		if err != nil {
+			return errors.Wrap(err, "Error building bundles")
+		}
+	}
+	return nil
+}
+
+func excludeName(man *swupd.Manifest, exclude string) {
+	for i := range man.Files {
+		if man.Files[i].Name == exclude {
+			man.Files = append(man.Files[:i], man.Files[i+1:]...)
+			break
+		}
+	}
+}
+
+func mergeMoMs(mixWS string, mixVer, lastVer int) error {
+	upstreamMoM, err := swupd.ParseManifestFile(filepath.Join(mixWS, "Manifest.MoM"))
+	if err != nil {
+		return err
+	}
+
+	mixerMoM, err := swupd.ParseManifestFile(
+		filepath.Join(mixWS, fmt.Sprintf("update/www/%d/Manifest.MoM", mixVer)))
+	if err != nil {
+		return err
+	}
+
+	// add mixerMoM filecount minus os-core
+	upstreamMoM.Header.FileCount += mixerMoM.Header.FileCount - 1
+	// need to set previous version, without this mixer version is the currently built version
+	if lastVer == 0 {
+		upstreamMoM.Header.Previous = upstreamMoM.Header.Version
+	} else {
+		upstreamMoM.Header.Previous = uint32(mixVer - 10)
+	}
+	// format is 1 until auto-format-bump support
+	upstreamMoM.Header.Format = 1
+	// system now on mixVer instead of upstream version
+	upstreamMoM.Header.Version = uint32(mixVer)
+	// remove os-core entry from upstreamMoM, we will replace with ours
+	excludeName(upstreamMoM, "os-core")
+	for i := range mixerMoM.Files {
+		if mixerMoM.Files[i].Name == "os-core-update-index" {
+			continue
+		}
+		mixerMoM.Files[i].Rename = swupd.MixManifest
+		upstreamMoM.Files = append(upstreamMoM.Files, mixerMoM.Files[i])
+	}
+
+	return upstreamMoM.WriteManifestFile("Manifest.MoM")
+}
+
+func main() {
+	var err error
+
+	flag.Parse()
+	if len(flag.Args()) != 2 {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	args := flag.Args()
+	bundle := args[0]
+	pkg := args[1]
+
+	ver, err := getCurrentVersion()
+	failIfErr(err)
+	mixVer := ver * 1000
+	if _, err = os.Stat(filepath.Join(mixWS, "builder.conf")); os.IsNotExist(err) {
+		err = setUpMixDir(bundle, ver, mixVer)
+		failIfErr(err)
+	}
+
+	err = os.Chdir(mixWS)
+	failIfErr(err)
+
+	b, err := builder.NewFromConfig(filepath.Join(mixWS, "builder.conf"))
+	failIfErr(err)
+	err = b.InitMix(fmt.Sprintf("%d", ver), fmt.Sprintf("%d", mixVer),
+		false, false, "https://download.clearlinux.org", false)
+	failIfErr(err)
+	b.NumBundleWorkers = runtime.NumCPU()
+	b.NumFullfileWorkers = runtime.NumCPU()
```

Comment:
At some stage probably ought to have parameters to set these numbers of workers.

Reply:
The parameters exist through the CLI, but not through the API unfortunately. I agree with you.